File: src/schema.ts
// JSON data contract between the extraction CLI (src/extract/*) and the SPA
// (app/*). Two documents: a cheap Manifest (always regenerated) and one
// expensive ConfigData blob per nixos/darwin configuration (extracted on
// demand). storePath is the universal join key: FileEntry.storePath matches
// the file strings in OptionEntry declarations/definitions.

export const SCHEMA_VERSION = 1

export interface Manifest {
  version: typeof SCHEMA_VERSION
  generatedAt: string // ISO timestamp
  extractor: string // extraction-code fingerprint at generation time (extract/fingerprint.ts)
  flake: FlakeInfo
  /** Output tree from `nix flake show --json`, normalized. */
  outputs: OutputNode
  /** Root-level inputs, keyed by input name. */
  inputs: Record<string, InputInfo>
  files: FileEntry[]
  /** file→file static import graph (self files only). */
  importEdges: ImportEdge[]
  /** Self files whose source references `inputs.<name>` (regex scan, like importEdges). */
  inputRefs: InputRef[]
  configurations: ConfigRef[]
  /** Derivation-typed outputs: packages, devShells, checks, formatter (see PackageRef). */
  packages: PackageRef[]
  /** Outputs that extend an input's same-named namespace (lib = nixpkgs.lib.extend …). */
  grafts: GraftInfo[]
  /** Top-level attr names per output — fills in where nix flake show says "unknown". */
  outputNames: Record<string, string[]>
  /** Non-fatal extraction problems, surfaced in the UI. */
  warnings: string[]
}

/**
 * A top-level output detected as a graft onto an input's same-named
 * namespace: ≥90% of the input's attr names reappear in the output. The UI
 * shows only `added` and hides the inherited bulk.
 */
export interface GraftInfo {
  /** Top-level output name, e.g. "lib". */
  output: string
  /** Direct input whose namespace it extends, e.g. "nixpkgs". */
  input: string
  /** Keys the flake adds on top of the input's namespace. */
  added: string[]
  /** Count of keys inherited (by name) from the input. */
  inherited: number
}

export interface FlakeInfo {
  /** Flake reference as given on the CLI. */
  ref: string
  /** self.outPath — the flake's own store path. */
  path: string
  description?: string
  rev?: string
  narHash?: string
}

export type OutputNode =
  | { kind: "attrset"; children: Record<string, OutputNode> }
  | { kind: "leaf"; type: string; name?: string; description?: string }
  /** `nix flake show` omitted this (other-system or eval failure). */
  | { kind: "omitted" }
  /** `nix flake show` could not classify it ({"unknown": true}). */
  | { kind: "unknown" }

export interface InputInfo {
  /**
   * Display name: the input attr for direct inputs, "parent/child" for
   * transitive ones (deduped — a followed input appears once).
   */
  name: string
  /**
   * flake.lock node key (differs from name for deduped/followed nodes).
   * Not rendered directly, but it is the join target `follows` points at.
   */
  nodeKey: string
  /** Present on inputs-of-inputs; the UI legend shows direct inputs only. */
  transitive?: true
  /**
   * Other ROOT-level input names that follow this same lock node (e.g.
   * "stable" for `inputs.stable.follows = "nixpkgs"`). The entry keeps the
   * real (non-follows) input's name; aliases are sorted for determinism.
   */
  aliases?: string[]
  type: string
  url?: string
  ref?: string
  rev?: string
  narHash?: string
  lastModified?: number
  /** toString inputs.<name>.outPath — prefix-match key for file attribution. */
  storePath?: string
  /** Present when this input follows another (value = followed node key). */
  follows?: string
}

export type FileOrigin =
  | { kind: "self" }
  /** patched: the file lives in a patched COPY of the input's source tree. */
  | { kind: "input"; input: string; patched?: true }
  /** group: display bucket for unattributed store roots ("source@abc1234"). */
  | { kind: "unknown"; group?: string }

/**
 * FileEntry.id codec. The format ("self:<rel>" | "input:<name>:<rel>") is a
 * client-server protocol — serve's /data/file/<id> route re-derives input
 * files from the id — so every construction and parse site goes through
 * these helpers. (resolveFile's "unknown:…"/"inline" buckets are app-internal
 * and opaque: parseFileId returns null for them.)
 */
export function makeFileId(
  origin: { kind: "self" } | { kind: "input"; input: string },
  relPath: string,
): string {
  return origin.kind === "self" ? `self:${relPath}` : `input:${origin.input}:${relPath}`
}

export type ParsedFileId =
  | { kind: "self"; relPath: string }
  | { kind: "input"; input: string; relPath: string }

export function parseFileId(id: string): ParsedFileId | null {
  if (id.startsWith("self:")) return { kind: "self", relPath: id.slice(5) }
  const m = id.match(/^input:([^:]+):(.+)$/)
  return m ? { kind: "input", input: m[1]!, relPath: m[2]! } : null
}

/** Display label for a file id: its relPath; opaque (unknown-bucket) ids as-is. */
export function displayLabel(id: string): string {
  return parseFileId(id)?.relPath ?? id
}

export interface FileEntry {
  /** Stable id — see makeFileId/parseFileId above. */
  id: string
  /** Path relative to its origin root. */
  relPath: string
  origin: FileOrigin
  /** Absolute /nix/store/... path — JOIN KEY to option file references. */
  storePath: string
  /** Last commit touching this file; self files with a local checkout only. */
  git?: { commit: string; date: string; subject: string }
}

// ---------------------------------------------------------------------------
// GET /data/file/<id> response (src/extract/highlight.ts tokenizes server-side)

export interface FileSource {
  text: string
  /** Flat, non-overlapping tree-sitter highlight spans over `text`; [] if tokenizing failed. */
  tokens: TokenRun[]
}

export interface TokenRun {
  start: number
  end: number
  /** Highlight-query capture name, e.g. "keyword", "string", "string.special.path". */
  name: string
}

/** Directed edge: `from` imports `to` (both FileEntry.id). */
export interface ImportEdge {
  from: string
  to: string
}

/** A self file whose source text references `inputs.<input>` (or `inputs'.<input>`). */
export interface InputRef {
  /** FileEntry.id of the referencing file. */
  file: string
  /** Canonical root input name (follows aliases resolved to the real input). */
  input: string
}

export type ConfigKind = "nixos" | "darwin"

export interface ConfigRef {
  /** "nixos/nebula", "darwin/k". */
  id: string
  kind: ConfigKind
  name: string
  /** Data file relative to the data dir, e.g. "config/nixos.nebula.json". */
  dataFile: string
  status: "pending" | "ok" | "error"
  error?: string
  extractedAt?: string
  optionCount?: number
  durationMs?: number
}

// ---------------------------------------------------------------------------
// Derivation-typed outputs: packages, devShells, checks, formatter. Extracted
// on demand (data/package/<safe-path>.json), same lifecycle as ConfigRef —
// enumerated for free from the already-normalized `outputs` tree (no extra
// eval), unlike ConfigRef which comes from a dedicated nix-side eval.

export interface PackageRef {
  /** path.join("/"), e.g. "packages/x86_64-linux/rtk". */
  id: string
  /** Output-tree path segments, e.g. ["packages", "x86_64-linux", "rtk"]. */
  path: string[]
  /** Data file relative to the data dir, e.g. "package/packages.x86_64-linux.rtk.json". */
  dataFile: string
  status: "pending" | "ok" | "error"
  error?: string
  extractedAt?: string
  durationMs?: number
}

export type BuilderKind =
  | "rustPlatform"
  | "buildGoModule"
  | "node"
  | "trivial"
  | "stdenv"
  | "unknown"

export interface PackageLicense {
  shortName?: string
  fullName?: string
  spdxId?: string
  url?: string
  free?: boolean
}

export interface PackageMaintainer {
  name?: string
  github?: string
  email?: string
}

export interface PackageMeta {
  description?: string
  homepage?: string
  /** Always normalized to a list, whatever shape meta.license took (attr/list/string). */
  license?: PackageLicense[]
  platforms?: string[]
  mainProgram?: string
  /** Capped — some nixpkgs packages list dozens. */
  maintainers?: PackageMaintainer[]
  /** meta.position, "file:line" — resolvable to a file chip when under the flake's own path. */
  position?: string
  broken?: boolean
  unfree?: boolean
}

export interface PackageSrc {
  storePath?: string
  url?: string
  rev?: string
  outputHash?: string
}

export interface PackageDeps {
  nativeBuildInputs: string[]
  buildInputs: string[]
  propagatedBuildInputs: string[]
}

/** One derivation phase script, e.g. {name: "buildPhase", script: "..."}. */
export interface DrvPhase {
  name: string
  script: string
  /** Flat, non-overlapping tree-sitter (bash) highlight spans over `script`; [] if tokenizing failed. */
  tokens: TokenRun[]
}

export interface DrvInputRef {
  drvPath: string
  name: string
  outputs: string[]
}

/**
 * `nix derivation show` output (instantiation only — never builds). Absent
 * when instantiation itself fails (recorded as a warning instead).
 */
export interface DrvInfo {
  drvPath: string
  system: string
  builderPath: string
  inputDrvs: DrvInputRef[]
  /** Scripts capped ~4000 chars each. */
  phases: DrvPhase[]
  doCheck?: boolean
  strictDeps?: boolean
  structuredAttrs?: boolean
}

/**
 * `nix path-info` for one output — present only when that output's path is
 * already valid in the local store (query-only; this tool never builds).
 */
export interface RuntimeInfo {
  outPath: string
  references: string[]
  narSize?: number
  closureSize?: number
}

// ---------------------------------------------------------------------------
// Per-package blob (data/package/<safe-path>.json)

export interface PackageData {
  /** SCHEMA_VERSION at extraction time — the SPA rejects mismatched blobs. */
  version: typeof SCHEMA_VERSION
  id: string
  path: string[]
  name?: string
  pname?: string
  /**
   * The derivation's own `version` attr — named pkgVersion (not `version`)
   * to avoid colliding with the schema-version discriminant above.
   */
  pkgVersion?: string
  builder: BuilderKind
  stdenv?: string
  system?: string
  meta?: PackageMeta
  src?: PackageSrc
  outputs: { name: string; outPath?: string }[]
  deps: PackageDeps
  drv?: DrvInfo
  /** Keyed by output name, e.g. runtime.out. */
  runtime?: Record<string, RuntimeInfo>
  warnings: string[]
}

// ---------------------------------------------------------------------------
// Per-configuration blob (data/config/<kind>.<name>.json)

export interface ConfigData {
  /** SCHEMA_VERSION at extraction time — the SPA rejects mismatched blobs. */
  version: typeof SCHEMA_VERSION
  id: string
  options: OptionEntry[]
  /**
   * storePath (or "<unknown-file>") -> indices into `options`, split by role.
   * Precomputed so the SPA never scans thousands of options per click:
   * `defines` = files that set a value, `declares` = files declaring the option.
   */
  fileIndex: Record<string, FileOptionRefs>
}

export interface FileOptionRefs {
  defines: number[]
  declares: number[]
}

/** Well-known mk-priority values (lib.mkOverride n). */
export const PRIO = {
  mkForce: 50,
  plain: 100,
  mkDefault: 1000,
  /** The option's own declared default (lib.mkOptionDefault). */
  optionDefault: 1500,
} as const

export interface OptionEntry {
  /** Option path, e.g. ["services", "nginx", "enable"]. */
  loc: string[]
  /** type.description, e.g. "boolean" or "string matching ...". */
  type?: string
  description?: string
  readOnly: boolean
  isDefined: boolean
  /** Winning definition priority; absent when !isDefined. */
  highestPrio?: number
  /** isDefined && highestPrio < 1500 — a real definition beat the default. */
  customized: boolean
  /** Merged value (scrubbed); absent for package-typed or throwing values. */
  value?: unknown
  /** True when evaluating/serializing the value threw. */
  valueError?: true
  /** True when the extractor skipped the value (package-typed, or a degraded chunk). */
  valueSkipped?: true
  default?: unknown
  defaultText?: string
  /** Declaring files. */
  declarations: DeclarationRef[]
  /** One entry per definition (options.<x>.definitionsWithLocations). */
  definitions: DefinitionRef[]
}

export interface DeclarationRef {
  file: string // store path or "<unknown-file>"
  /** 1-based declaration site (option.declarationPositions); absent on older module systems. */
  line?: number
  column?: number
}

export interface DefinitionRef {
  file: string // store path or "<unknown-file>"
  value?: unknown
  valueError?: true
  /** True when the extractor skipped this definition's value. */
  valueSkipped?: true
  /**
   * mkOverride priority lifted from a wrapper surviving in this definition's
   * raw value. The real module system strips wrappers AND drops losing-prio
   * definitions before exposing definitionsWithLocations (lib filterOverrides),
   * so there absent means "merged at the option's highestPrio"; it is present
   * only on module systems that expose raw definition values.
   */
  prio?: number
}

/** Sentinel file string the module system uses for inline/anonymous modules. */
export const UNKNOWN_FILE = "<unknown-file>"

File: rust/src/schema.rs
// JSON data contract between the extractor and the SPA — a 1:1 port of
// src/schema.ts. Field names, tag values, and optional-field omission must
// match the TypeScript output byte-for-byte in shape: the Svelte app consumes
// these documents unchanged.

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const SCHEMA_VERSION: u32 = 1;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Manifest {
    pub version: u32,
    pub generated_at: String,
    pub extractor: String,
    pub flake: FlakeInfo,
    pub outputs: OutputNode,
    pub inputs: IndexMap<String, InputInfo>,
    pub files: Vec<FileEntry>,
    pub import_edges: Vec<ImportEdge>,
    pub input_refs: Vec<InputRef>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub overlay_defs: Option<Vec<OverlayDef>>,
    pub input_follows: Vec<InputFollow>,
    pub configurations: Vec<ConfigRef>,
    pub packages: Vec<PackageRef>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub package_reverse_deps: Option<IndexMap<String, Vec<String>>>,
    pub grafts: Vec<GraftInfo>,
    pub output_names: IndexMap<String, Vec<String>>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GraftInfo {
    pub output: String,
    pub input: String,
    pub added: Vec<String>,
    pub inherited: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlakeInfo {
    pub r#ref: String,
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rev: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nar_hash: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum OutputNode {
    #[serde(rename = "attrset")]
    Attrset { children: IndexMap<String, OutputNode> },
    #[serde(rename = "leaf")]
    Leaf {
        r#type: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        name: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        description: Option<String>,
    },
    #[serde(rename = "omitted")]
    Omitted,
    #[serde(rename = "unknown")]
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InputInfo {
    pub name: String,
    pub node_key: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transitive: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub aliases: Option<Vec<String>>,
    pub r#type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rev: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nar_hash: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_modified: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub store_path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub follows: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum FileOrigin {
    #[serde(rename = "self")]
    SelfOrigin,
    #[serde(rename = "input")]
    Input {
        input: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        patched: Option<bool>,
    },
    #[serde(rename = "unknown")]
    Unknown {
        #[serde(skip_serializing_if = "Option::is_none")]
        group: Option<String>,
    },
}

/// FileEntry.id codec — "self:<rel>" | "input:<name>:<rel>" (client-server protocol).
pub fn make_file_id_self(rel_path: &str) -> String {
    format!("self:{rel_path}")
}

pub fn make_file_id_input(input: &str, rel_path: &str) -> String {
    format!("input:{input}:{rel_path}")
}

#[derive(Debug, Clone, PartialEq)]
pub enum ParsedFileId {
    SelfFile { rel_path: String },
    InputFile { input: String, rel_path: String },
}

pub fn parse_file_id(id: &str) -> Option<ParsedFileId> {
    if let Some(rel) = id.strip_prefix("self:") {
        return Some(ParsedFileId::SelfFile { rel_path: rel.to_string() });
    }
    let rest = id.strip_prefix("input:")?;
    let colon = rest.find(':')?;
    let (input, rel) = (&rest[..colon], &rest[colon + 1..]);
    if input.is_empty() || rel.is_empty() {
        return None;
    }
    Some(ParsedFileId::InputFile { input: input.to_string(), rel_path: rel.to_string() })
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileEntry {
    pub id: String,
    pub rel_path: String,
    pub origin: FileOrigin,
    pub store_path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub git: Option<GitFileInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitFileInfo {
    pub commit: String,
    pub date: String,
    pub subject: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileSource {
    pub text: String,
    pub tokens: Vec<TokenRun>,
}

/// Highlight span over the text — start/end are UTF-16 code-unit indices
/// (the client slices JS strings with them).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenRun {
    pub start: usize,
    pub end: usize,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportEdge {
    pub from: String,
    pub to: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InputRef {
    pub file: String,
    pub input: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OverlayDef {
    pub name: String,
    pub file: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attrs: Option<Vec<OverlayAttr>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OverlayAttr {
    pub name: String,
    pub kind: OverlayAttrKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OverlayAttrKind {
    Add,
    Override,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InputFollow {
    pub name: String,
    pub target: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ConfigKind {
    Nixos,
    Darwin,
}

impl ConfigKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            ConfigKind::Nixos => "nixos",
            ConfigKind::Darwin => "darwin",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RefStatus {
    Pending,
    Ok,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfigRef {
    pub id: String,
    pub kind: ConfigKind,
    pub name: String,
    pub data_file: String,
    pub status: RefStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extracted_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub option_count: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageRef {
    pub id: String,
    pub path: Vec<String>,
    pub data_file: String,
    pub status: RefStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extracted_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum BuilderKind {
    #[serde(rename = "rustPlatform")]
    RustPlatform,
    #[serde(rename = "buildGoModule")]
    BuildGoModule,
    #[serde(rename = "node")]
    Node,
    #[serde(rename = "trivial")]
    Trivial,
    #[serde(rename = "stdenv")]
    Stdenv,
    #[serde(rename = "unknown")]
    Unknown,
}

impl BuilderKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            BuilderKind::RustPlatform => "rustPlatform",
            BuilderKind::BuildGoModule => "buildGoModule",
            BuilderKind::Node => "node",
            BuilderKind::Trivial => "trivial",
            BuilderKind::Stdenv => "stdenv",
            BuilderKind::Unknown => "unknown",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageLicense {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub short_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub full_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub spdx_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub free: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageMaintainer {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub github: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageMeta {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub homepage: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub license: Option<Vec<PackageLicense>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub platforms: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub main_program: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub maintainers: Option<Vec<PackageMaintainer>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub broken: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unfree: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageSrc {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub store_path: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rev: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_hash: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageDeps {
    pub native_build_inputs: Vec<String>,
    pub build_inputs: Vec<String>,
    pub propagated_build_inputs: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DrvPhase {
    pub name: String,
    pub script: String,
    pub tokens: Vec<TokenRun>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DrvInputRef {
    pub drv_path: String,
    pub name: String,
    pub outputs: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DrvInfo {
    pub drv_path: String,
    pub system: String,
    pub builder_path: String,
    pub input_drvs: Vec<DrvInputRef>,
    pub phases: Vec<DrvPhase>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub do_check: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub strict_deps: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub structured_attrs: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RuntimeInfo {
    pub out_path: String,
    pub references: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nar_size: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub closure_size: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageOutput {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub out_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PackageData {
    pub version: u32,
    pub id: String,
    pub path: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pname: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pkg_version: Option<String>,
    pub builder: BuilderKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stdenv: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meta: Option<PackageMeta>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub src: Option<PackageSrc>,
    pub outputs: Vec<PackageOutput>,
    pub deps: PackageDeps,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub drv: Option<DrvInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub runtime: Option<IndexMap<String, RuntimeInfo>>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfigData {
    pub version: u32,
    pub id: String,
    pub options: Vec<OptionEntry>,
    pub file_index: IndexMap<String, FileOptionRefs>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileOptionRefs {
    pub defines: Vec<usize>,
    pub declares: Vec<usize>,
}

/// Well-known mk-priority values (lib.mkOverride n).
pub const PRIO_OPTION_DEFAULT: i64 = 1500;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OptionEntry {
    pub loc: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub read_only: bool,
    pub is_defined: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub highest_prio: Option<i64>,
    pub customized: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value_error: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value_skipped: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value_names: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_names: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_text: Option<String>,
    pub declarations: Vec<DeclarationRef>,
    pub definitions: Vec<DefinitionRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeclarationRef {
    pub file: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub line: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub column: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub via: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DefinitionRef {
    pub file: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value_error: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value_skipped: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value_names: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub via: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prio: Option<i64>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn file_id_round_trip() {
        assert_eq!(
            parse_file_id("self:modules/a.nix"),
            Some(ParsedFileId::SelfFile { rel_path: "modules/a.nix".into() })
        );
        assert_eq!(
            parse_file_id("input:nixpkgs:lib/default.nix"),
            Some(ParsedFileId::InputFile {
                input: "nixpkgs".into(),
                rel_path: "lib/default.nix".into()
            })
        );
        assert_eq!(parse_file_id("unknown:whatever"), None);
        assert_eq!(parse_file_id("inline"), None);
    }

    #[test]
    fn output_node_tags() {
        let leaf = OutputNode::Leaf { r#type: "derivation".into(), name: None, description: None };
        assert_eq!(serde_json::to_string(&leaf).unwrap(), r#"{"kind":"leaf","type":"derivation"}"#);
        let omitted = OutputNode::Omitted;
        assert_eq!(serde_json::to_string(&omitted).unwrap(), r#"{"kind":"omitted"}"#);
    }

    #[test]
    fn origin_tags() {
        assert_eq!(
            serde_json::to_string(&FileOrigin::SelfOrigin).unwrap(),
            r#"{"kind":"self"}"#
        );
        assert_eq!(
            serde_json::to_string(&FileOrigin::Input { input: "nixpkgs".into(), patched: None })
                .unwrap(),
            r#"{"kind":"input","input":"nixpkgs"}"#
        );
    }
}
